### PR TITLE
Iaco 678 sdk to return user data to merchant

### PR DESCRIPTION
**Pull request by yuval-netanel**

### CODE REVIEW

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java [hunk 1]
```diff
@@ -16,6 +18,10 @@
      * This item uniquely identifies the user at Klarna.
      */
     public static final String EXTRA_USER_TOKEN = "userToken";
```
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java]: Is this still being used?

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java [hunk 2]
```diff
@@ -24,9 +30,19 @@ protected String getUrl() {
 
     @Override
     protected void handleUserReadyEvent(Map<Object, Object> payload) {
-        String token = (String) payload.get("userToken");
         Intent result = new Intent();
-        result.putExtra(EXTRA_USER_TOKEN, token);
+
+        // Backwards compatibility.
```
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java]: Do we have anyone to be backwards compatible with at this point in time?

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java [hunk 3]
```diff
@@ -24,9 +30,19 @@ protected String getUrl() {
...
+        result.putExtra(EXTRA_USER_TOKEN, (String)payload.get(PAYLOAD_USER_TOKEN));
+
+        Map<?, ?> userDetails =  payload.containsKey(PAYLOAD_USER_DETAILS) ? (Map<?, ?>)payload.get(PAYLOAD_USER_DETAILS) : (Map<?, ?>)Collections.EMPTY_MAP;
+        RegistrationResult registrationResult =new RegistrationResult(
```
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java]: `=new` -> `= new`

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java (hunk 2 above)
Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java]: This _does_ allow us not to change the docs, but that is not a good around reason to hang on to this.

**Comment by grozen** [comment]
Approved!

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java (hunk 2 above)
Comment (yuval-netanel) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java]: Removing this

Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java]: +1

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java (hunk 3 above)
Comment (yuval-netanel) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java]: :+1

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java [hunk 4]
```diff
@@ -24,9 +30,19 @@ protected String getUrl() {
...
     @Override
     protected void handleUserReadyEvent(Map<Object, Object> payload) {
-        String token = (String) payload.get("userToken");
         Intent result = new Intent();
-        result.putExtra(EXTRA_USER_TOKEN, token);
+
+        // Backwards compatibility.
+        result.putExtra(EXTRA_USER_TOKEN, (String)payload.get(PAYLOAD_USER_TOKEN));
+
+        Map<?, ?> userDetails =  payload.containsKey(PAYLOAD_USER_DETAILS) ? (Map<?, ?>)payload.get(PAYLOAD_USER_DETAILS) : (Map<?, ?>)Collections.EMPTY_MAP;
```
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java]: No need for the default value.
Payload is designed to always contain a HASH of 'userDetails'.

Comment (yuval-netanel) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationActivity.java]: OK

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationResult.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+        RegistrationResult that = (RegistrationResult) o;
+
+        if (phoneNumber != null ? !phoneNumber.equals(that.phoneNumber) : that.phoneNumber != null)
+            return false;
+        if (token != null ? !token.equals(that.token) : that.token != null) return false;
+        if (userDetails != null ? !userDetails.equals(that.userDetails) : that.userDetails != null)
+            return false;
+
+        return true;
```
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationResult.java]: just suggesting:
```
        if (o == this)
            return true;

        if (!(o instanceof RegistrationResult))
            return false;

        RegistrationResult that = (RegistrationResult) o;

        if(phoneNumber == null) {
            return that.phoneNumber == null;
        }
        if(token == null) {
            return that.token == null;
        }
        if(userDetails == null) {
            return that.userDetails == null;
        }
        return phoneNumber.equals(that.phoneNumber) &&
               token.equals(that.token) &&
               userDetails.equals(that.userDetails);
```

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationResult.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        if (o == null || getClass() != o.getClass()) return false;
+
+        RegistrationResult that = (RegistrationResult) o;
+
+        if (phoneNumber != null ? !phoneNumber.equals(that.phoneNumber) : that.phoneNumber != null)
+            return false;
+        if (token != null ? !token.equals(that.token) : that.token != null) return false;
+        if (userDetails != null ? !userDetails.equals(that.userDetails) : that.userDetails != null)
+            return false;
+
```
Comment (tomerlevi444) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationResult.java:55]: just suggesting:

```
        if (o == this)
            return true;

        if (!(o instanceof RegistrationResult))
            return false;

        RegistrationResult that = (RegistrationResult) o;

        if(phoneNumber == null && that.phoneNumber != null) {
            return false;
        }
        if(token == null && that.token != null) {
            return false;
        }
        if(userDetails == null && that.userDetails != null) {
            return false;
        }

        return phoneNumber.equals(that.phoneNumber) &&
               token.equals(that.token) &&
               userDetails.equals(that.userDetails);
```

Comment (grozen) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationResult.java:55]: I agree this is nicer to look at.
I don't agree that `if`s without a block after them is something that should be considered legal.

File: klarna-on-demand/src/test/java/com/klarna/ondemand/RegistrationActivityTest.java [hunk 7]
```diff
@@ -58,12 +52,39 @@ public void beforeEach() {
...
+
+        expectedIntent.putExtra(RegistrationActivity.EXTRA_REGISTRATION_RESULT,
+                new RegistrationResult("my_token", "my_phoneNumber", userDetails));
+
+        verify(registrationActivity).setResult(eq(Activity.RESULT_OK), eq(expectedIntent));
+        verify(registrationActivity).finish();
+    }
+
+    @Test
+    public void handleUserReadyEvent_ShouldPutEmptyUserDetailsInTheRegistrationResult_WhenPaylodDoesNotContainUserDetails() {
```
Comment (tomerlevi444) [inline klarna-on-demand/src/test/java/com/klarna/ondemand/RegistrationActivityTest.java]: This test is not relevant

**Comment by tomerlevi444** [comment]
Approved

File: klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationResult.java (hunk 6 above)
Comment (yuval-netanel) [inline klarna-on-demand/src/main/java/com/klarna/ondemand/RegistrationResult.java:55]: Generally I agree
But this is auto generated code, so I will leave it as is

File: example/src/main/java/com/klarna/example/MainActivity.java [hunk 8]
```diff
@@ -29,6 +30,7 @@
     public static final int REGISTRATION_REQUEST_CODE = 1;
     public static final int PREFERENCES_REQUEST_CODE = 2;
     private static final String USER_TOKEN_KEY = "userToken";
+    private static final String LOG_TAG = MainActivity.class.getName();
```
Comment (MaTriXy) [inline example/src/main/java/com/klarna/example/MainActivity.java:33]: Just For Future knowledge - if you run Proguard and don't keep classes names you can end up with  a TAG like : "aaa"
i know it's not the case but keep that in mind for future :)

**Pull request merged**